### PR TITLE
Cleanup: replace multiple HashMap with STL unordered_map

### PR DESCRIPTION
**Pull request by chitianhao**
This is not related to TCL but we are also trying to get rid of `Map.h`.
It's interesting that we have so many different kinds of hash tables in different places.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
Yes, one of the goals of this project is to reduce the number of hash map classes in use.

File: iocore/net/P_SNIActionPerformer.h [hunk 1]
```diff
@@ -37,7 +37,7 @@
 #include "P_SSLNextProtocolAccept.h"
 #include "tscore/ink_inet.h"
 
-extern Map<int, SSLNextProtocolSet *> snpsMap;
+extern std::unordered_map<int, SSLNextProtocolSet *> snpsMap;
```
Comment (ywkaras) [inline iocore/net/P_SNIActionPerformer.h]: Why don't you need #include <unordered_map> ?

File: iocore/net/SSLSNIConfig.cc [hunk 2]
```diff
@@ -112,29 +114,25 @@ SNIConfigParams::SNIConfigParams() {}
...
 {
-  auto actionVec = sni_action_map.get(servername);
-  if (!actionVec) {
-    Vec<cchar *> keys;
-    wild_sni_action_map.get_keys(keys);
-    for (int i = 0; i < static_cast<int>(keys.length()); i++) {
+  auto action_it = sni_action_map.find(servername);
+  if (action_it != sni_action_map.end()) {
+    for (auto &&it : wild_sni_action_map) {
       std::string_view sv{servername, strlen(servername)};
```
Comment (SolidWallOfCode) [inline iocore/net/SSLSNIConfig.cc]: You don't need to do `strlen` explicitly, if it's a C-style string you can pass it as a `char const*`. E.g.
```
std::string_view sv{servername};
```

File: iocore/net/P_SNIActionPerformer.h (hunk 1 above)
Comment (chitianhao) [reply iocore/net/P_SNIActionPerformer.h]: It's included by some other file that this file includes. I'll add it anyway.

Comment (SolidWallOfCode) [reply iocore/net/P_SNIActionPerformer.h]: Yes. Never depend on "some other file" including what you need. Explicitly include what you need.

File: iocore/net/P_SNIActionPerformer.h [hunk 3]
```diff
@@ -67,8 +67,9 @@ class DisableH2 : public ActionItem
     auto ssl_vc     = reinterpret_cast<SSLNetVConnection *>(cont);
     auto accept_obj = ssl_vc ? ssl_vc->accept_object : nullptr;
     if (accept_obj && accept_obj->snpa && ssl_vc) {
-      auto nps = snpsMap.get(accept_obj->id);
-      ssl_vc->registerNextProtocolSet(reinterpret_cast<SSLNextProtocolSet *>(nps));
+      if (auto it = snpsMap.find(accept_obj->id); it != snpsMap.end()) {
```
Comment (zwoop) [inline iocore/net/P_SNIActionPerformer.h]: Here's this bloody pattern again... You make it look like a foor-loop, it's not an iterator!!

```
if (auto obj = snpsMap.find(accept_obj->id); obj != snpsMap.end()) {
```

**Comment by zwoop** [comment]
Hmmm, it seems we have done

```
if (auto it = ...
```

allo over the code... Why oh why ?

File: iocore/net/SSLSNIConfig.cc (hunk 2 above)
Comment (zwoop) [reply iocore/net/SSLSNIConfig.cc]: As seen in other PRs, should this now be 

```
     for (auto const &it: wild_sni_action_map)
```

?

File: iocore/net/SSLSNIConfig.cc [hunk 4]
```diff
@@ -38,19 +38,18 @@
...
 {
-  NextHopProperty *nps = nullptr;
-  nps                  = next_hop_table.get(servername);
-  if (!nps) {
-    nps = wild_next_hop_table.get(servername);
+  if (auto it = next_hop_table.find(servername); it != next_hop_table.end()) {
+    return it->second;
   }
-  return nps;
+  auto it = wild_next_hop_table.find(servername);
```
Comment (zwoop) [inline iocore/net/SSLSNIConfig.cc]: Why not be consistent on the pattern at least?

```

  if (auto it = wild_next_hop_table.find(servername); it != wild_next_hop_table.end()) {
    return it->second;
  } 

  return nullptr;
}
```

File: iocore/net/SSLSNIConfig.cc [hunk 5]
```diff
@@ -92,15 +91,17 @@ SNIConfigParams::loadSNIConfig()
...
-    nps->name                   = ats_strdup(servername);
-    nps->verifyServerPolicy     = item.verify_server_policy;
-    nps->verifyServerProperties = item.verify_server_properties;
-    nps->ctx                    = clientCTX;
-    if (wildcard) {
-      wild_next_hop_table.put(nps->name, nps);
-    } else {
-      next_hop_table.put(nps->name, nps);
+    if (servername) { // a safety check
+      NextHopProperty *nps        = new NextHopProperty();
```
Comment (zwoop) [inline iocore/net/SSLSNIConfig.cc]: This is me harping, but I really, really prefer that you put an empty line after local variable declaration. Not only does it look cleaner, but it really helps clang-format to not do dumb stuff such as aligning the variable declarations with the variable assignments

Comment (chitianhao) [reply iocore/net/SSLSNIConfig.cc]: Sure.

**Comment by chitianhao** [comment]
Updated. Should be ready to go @zwoop @SolidWallOfCode

**Comment by shinrich** [comment]
Requesting to be pulled into the 8.1.x branch so other requested PR's cherry-pick more directly.

**Pull request merged**